content/blog/2023/2023-09-16-create-openfeign-http-client.md

```diff
@@ -0,0 +1,503 @@
+
+```
+
+### Applying Circuit Breaker
+
+### Configuring Logger
+[SLF4JModule](https://github.com/OpenFeign/feign/blob/master/slf4j) is used to send Feign's logging to [SLF4J](http://www.slf4j.org/). With SLF4J, we can easily use a logging backend of our choice (Logback, Log4J, etc.)
+
+Here is an example about building the client:
+```java
+final CalculatorService target = Feign.builder()
+                                      .logger(new Slf4jLogger())
+                                      .logLevel(Level.FULL)
+                                      .target(CalculatorService.class, HOST);
+
+```
+To use SLF4J with Feign, add both the SLF4J module and an SLF4J binding of our choice to the classpath. 
+Then, configure Feign to use the Slf4jLogger:
+
+### Configuring Request Interceptors
+Request Interceptors in Feign allow us to **customize and manipulate HTTP requests before they are sent to the remote server**. They are useful for a variety of purposes, such as adding custom headers, logging, authentication, or request modification.
+
+Here's why we might want to use Request Interceptors in Feign:
+1. **Authentication**: We can use a Request Interceptor **to add authentication tokens or credentials to every request**. For example, adding an "Authorization" header with a JWT token.
+
+1. **Logging**: Interceptors are helpful **for logging incoming and outgoing requests and responses. This can be useful for debugging and monitoring**.
+
+1. **Request Modification**: We can **modify the request before it's sent**. This includes changing **headers, query parameters, or even the request body**.
+
+1. **Rate Limiting**: Implementing rate limiting by inspecting the number of requests being made and deciding whether to **allow or block a request**.
+
+1. **Caching**: Caching request/response data based on specific criteria.
+
+Here is a code snippet to demonstrate how to use request interception:
+```java
+static class AuthorizationInterceptor implements RequestInterceptor {
+  @Override public void apply(RequestTemplate template) {
+    // Check if token is present, if not, add it
+    template.header("Authorization", "Bearer " + generatedToken);
+  }
+}
+
+public class CalculatorServiceTest {
+  public static void main(String[] args) {
+    final interceptor = new AuthorizationInterceptor();
+    final CalculatorService target = Feign.builder()
+                                          .requestInterceptor(interceptor)
+                                          .target(CalculatorService.class, HOST);
+  }
+}
+
+```
+Implement `RequestInterceptor` and overrride its `apply` method. `apply` method would have custom logic as per requirements.
+
+### Configuring Retryer
+OpenFeign Retryer is a component that allows us to configure how Feign handles retries when a request fails. It can be particularly useful for handling transient failures in network communications. We can specify conditions under which Feign should automatically retry a failed request.
```

Comment:
Do I need to add a dependency to my project to use `Retryer`?

Reply:
No.. `Retryer` is included in feign-core